library/proc_macro/src/diagnostic.rs

```diff
@@ -7,42 +8,122 @@ use crate::Span;
+    fn span(self) -> Span
+    where
+        Self: Sized,
+    {
+        fn join_all(mut spans: impl Iterator<Item = Span>) -> Option<Span> {
+            let mut joined = spans.next()?;
+            for span in spans {
+                joined = match joined.join(span) {
+                    Some(joined) => joined,
+                    // Skips any bad joins.
+                    None => continue,
+                };
+            }
+
+            Some(joined)
+        }
+
+        join_all(self.spans()).unwrap_or_else(Span::call_site)
+    }
+
+    /// Create an error attached to the `Span`.
+    #[must_use = "diagnostics do nothing unless emitted"]
+    fn error(self, msg: &str) -> Diagnostic
+    where
+        Self: Sized,
+    {
+        Diagnostic::error(msg).mark(self.span())
+    }
+
+    // FIXME lint-associated warnings
+    // /// Create a warning attached to the `Span`.
+    // fn warning(self, lint: &Lint, msg: &str) -> Diagnostic {
+    //     Diagnostic::warning(lint, msg).mark(self.span())
+    // }
+
+    /// Create an note attached to the `Span`.
+    #[must_use = "diagnostics do nothing unless emitted"]
+    fn note(self, msg: &str) -> Diagnostic
+    where
+        Self: Sized,
+    {
+        Diagnostic::note(msg).mark(self.span())
+    }
 }
 
 #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-impl MultiSpan for Span {
-    fn into_spans(self) -> Vec<Span> {
-        vec![self]
+impl<S: Spanned, I: IntoIterator<Item = S>> Spanned for I {
+    // FIXME This _really_ should not be exposed publicly. It may be worthwhile
+    // to wait until `type_alias_impl_trait` is stabilized before exposing this
+    // type parameter.
+    //
+    // This is probably the case for other impls of this trait, but those are
+    // relatively trivial compared to this.
+    type Iter = iter::FlatMap<I::IntoIter, S::Iter, fn(S) -> S::Iter>;
+
+    fn spans(self) -> Self::Iter {
+        self.into_iter().flat_map(Spanned::spans)
     }
 }
 
 #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-impl MultiSpan for Vec<Span> {
-    fn into_spans(self) -> Vec<Span> {
+impl Spanned for Span {
+    type Iter = iter::Once<Self>;
+
+    fn spans(self) -> Self::Iter {
+        iter::once(self)
+    }
+
+    fn span(self) -> Span {
         self
     }
 }
 
-#[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
-impl<'a> MultiSpan for &'a [Span] {
-    fn into_spans(self) -> Vec<Span> {
-        self.to_vec()
-    }
+macro_rules! impl_span_passthrough {
+    ($($type:ty)*) => {$(
+        #[unstable(feature = "proc_macro_diagnostic", issue = "54140")]
+        impl Spanned for $type {
+            type Iter = iter::Once<Span>;
+
+            fn spans(self) -> Self::Iter {
+                iter::once(self.span())
+            }
+
+            fn span(self) -> Span {
+                Self::span(&self)
+            }
+        }
+    )*}
 }
 
+impl_span_passthrough![
```

Comment:
Is this known as passthrough or delegate?

Reply:
To me delegation is this in reverse. Doesn't much matter to me either way, though.